### PR TITLE
Make sure mobile understands community share link by Desktop

### LINKED ISSUE #14560: "Not found" screen when following a link to a community created on Desktop
# Bug Report

## Problem 

If a user follows a link shared from a community created on desktop they are redirected to the `Not found / Fetch community` screen.
The reason may be that community keys in deep links are incompatible between desktop and mobile.
We should also support compressed keys (`zQJ4nf...` format). 

It works fine if you use the `Invite new people` option (`Overview` > `Invite new people` > select contacts > `Next` > `Send inites`) . 

### Reproduction
1. User A (desktop) adds User B (mobile) to contacts 
2. User A: create a community on desktop
3. User A: go to `Overview` > `Share community` > Copy the link
Screenshot 2022-12-15 at 18 12 10

4. User A: paste the link in 1-1 chat with User B and send the message
5. User B: follow the link

https://user-images.githubusercontent.com/67952253/207925594-69e8bcbe-df91-4fa1-89bd-fa12a65ce978.mp4

#### Expected behavior
- The link is resolved in the 1-1 chat view as an invitation link (with the `View` button)
- Redirect to the community when following the link

#### Actual behavior
- The link is not resolved
- Redirect to the `Not found` screen

### Additional Information

- Status version: nightly 15/12, desktop 0.8.1
- Operating System: Android, iOS

### Note
Please, ping me for creating a separate issue for compressed keys support if needed.

### PR DESCRIPTION
**Pull request by siddarthkay**
fixes #14560

### Summary

When a community is created on Desktop its share url looks like : `https://join.status.im/c/zQ3shTAten2v9CwyQD1Kc7VXAqNPDcHZAMsfbLHCZEx6nFqk9`

When a community is created on Mobile its share url looks like : 
`https://join.status.im/c/0x025596a7ff87da36860a84b0908191ce60a504afc94aac93c1abd774f182967ce6`

This PR attempts to translate the received community ID ( when you consider the above example )
from 
`zQ3shTAten2v9CwyQD1Kc7VXAqNPDcHZAMsfbLHCZEx6nFqk9`
to 
`0x025596a7ff87da36860a84b0908191ce60a504afc94aac93c1abd774f182967ce6`

So that mobile can access community info from mailserver and it can become usable on mobile.

We should ultimately move communities to only support the shorter key format on mobile as well, 
now that I have the necessary research in place, I will do that in a separate PR.

#### Platforms
- Android
- iOS

##### Functional

- 1-1 chats
- public chats
- group chats
-communities

### Steps to test

- Open Status
- Go to Chat
- Paste share link received from desktop into a chat with anyone
- You should should be able to navigate to the community on tap 

status: ready

### CODE REVIEW

File: status-go-version.json [hunk 1]
```diff
@@ -3,7 +3,7 @@
     "_comment": "Instead use: scripts/update-status-go.sh <rev>",
     "owner": "status-im",
     "repo": "status-go",
-    "version": "v0.117.1a",
-    "commit-sha1": "b4bdfd3df6cf5fb91ab2d0e9f3b38e8d1b9703e5",
-    "src-sha256": "1s50q53p1j5ysnbj88c52zb2rh1ms21pnf6bjc7wja8gzcpih2wl"
+    "version": "14560-support-compressed-keys-for-communities",
```
Comment (siddarthkay) [inline status-go-version.json]: status-go version is my branch because that PR can not be merged in due to some CI issues.
I will update this to point to develop once that is fixed.
This is just to ensure that builds work fine and QA can test the fix.

File: src/status_im/chat/models/link_preview.cljs [hunk 2]
```diff
@@ -62,6 +62,7 @@
 (rf/defn resolve-community-info
   {:events [::resolve-community-info]}
   [{:keys [db]} community-id]
+  (log/info "resolve-community-info with community-id " community-id)
```
Comment (siddarthkay) [inline src/status_im/chat/models/link_preview.cljs]: should remove useless log here

File: src/status_im/native_module/core.cljs [hunk 3]
```diff
@@ -278,6 +278,52 @@
...
+  (log/info "[native-module] Compressing public key"
+            {:fn          :convert-desktop-key-to-mobile-key
+             :desktop-key desktop-key})
+  (.deserializeAndCompressKey ^js (status) desktop-key callback))
+
+
+(defn serialise-public-key-into-compressed-key
+  "Provides public key to status-go and gets back a compressed key"
+  [public-key callback]
+  (let [serialisation-key "z"
```
Comment (siddarthkay) [inline src/status_im/native_module/core.cljs]: will move "z" "0xe701" to a constants file and add a docstring

File: src/status_im/native_module/core.cljs [hunk 4]
```diff
@@ -278,6 +278,52 @@
...
+    (log/info "[native-module] Compressing public key"
+              {:fn             :serialise-public-key-into-compressed-key
+               :public-key     public-key
+               :multi-code-key multi-code-key})
+    (.multiformatSerializePublicKey ^js (status) multi-code-key serialisation-key callback)))
+
+(defn deserialise-compressed-key-into-public-key
+  "Provides compressed key to status-go and gets back the uncompressed public key"
+  [public-key callback]
+  (let [deserialisation-key "f"]
```
Comment (siddarthkay) [inline src/status_im/native_module/core.cljs]: will move "f" to a constants file and add a docstring

File: src/status_im/utils/universal_links/core.cljs [hunk 5]
```diff
@@ -76,7 +77,24 @@
 (rf/defn handle-community
   [cofx {:keys [community-id]}]
   (log/info "universal-links: handling community" community-id)
-  (navigation/navigate-to-cofx cofx :community {:community-id community-id}))
+  (navigation/navigate-to-cofx cofx :community {:community-id community-id})
+)
+
+(rf/defn handle-navigation-to-desktop-community-from-mobile
+  {:events [:handle-navigation-to-desktop-community-from-mobile]}
+  [{:keys [db]} cofx de-serialised-key]
```
Comment (siddarthkay) [inline src/status_im/utils/universal_links/core.cljs:85]: get rid of {:keys [db]} , its not used inside the function body

File: src/status_im/utils/universal_links/core.cljs [hunk 6]
```diff
@@ -76,7 +77,24 @@
...
+  (navigation/navigate-to-cofx cofx :community {:community-id de-serialised-key})
+)
+
+(rf/defn handle-desktop-community
+  [cofx {:keys [community-id]}]
+  (status/deserialize-and-compress-key
+   community-id
+   (fn [de-serialised-key]
```
Comment (ilmotta) [inline src/status_im/utils/universal_links/core.cljs]: I think we should try to be as consistent as possible with naming. This PR is using the word `deserialize` in multiple ways, sometimes using the British convention with `s`, sometimes with a dash. I suggest we stick to only one way. The dash seems unnecessary though.

File: src/status_im/ui/screens/chat/message/link_preview.cljs [hunk 7]
```diff
@@ -19,9 +19,14 @@
...
     :else                                                  false))
 
+;;siddarthkay : todo add code for previews here once previews are fixed in new messaging ui
+
 (defn community-id-from-link
   [link]
-  (nth (re-find constants/regx-community-universal-link link) 4))
+  (let [community-id (nth (re-find constants/regx-community-universal-link link) 4)]
+    community-id
+  )
```
Comment (ilmotta) [inline src/status_im/ui/screens/chat/message/link_preview.cljs]: Please, remove all dangling parentheses in this PR.

File: src/status_im/ui/screens/chat/message/link_preview.cljs [hunk 8]
```diff
@@ -19,9 +19,14 @@
     (string/starts-with? link (str "https://www." domain)) true
     :else                                                  false))
 
+;;siddarthkay : todo add code for previews here once previews are fixed in new messaging ui
+
 (defn community-id-from-link
   [link]
-  (nth (re-find constants/regx-community-universal-link link) 4))
+  (let [community-id (nth (re-find constants/regx-community-universal-link link) 4)]
```
Comment (ilmotta) [inline src/status_im/ui/screens/chat/message/link_preview.cljs]: Why the changes in this function? It seems to unnecessarily store in the community ID in a binding.

File: src/status_im/ui/screens/chat/message/link_preview.cljs [hunk 9]
```diff
@@ -19,9 +19,14 @@
     (string/starts-with? link (str "https://www." domain)) true
     :else                                                  false))
 
+;;siddarthkay : todo add code for previews here once previews are fixed in new messaging ui
```
Comment (ilmotta) [inline src/status_im/ui/screens/chat/message/link_preview.cljs]: This TODO could be removed and a Github issue created instead.

File: src/status_im/router/core.cljs [hunk 10]
```diff
@@ -229,7 +229,11 @@
       (cb {:type handler :community-id (:community-id route-params)})
 
       (= handler :community)
-      (cb {:type handler :community-id (:community-id route-params)})
+      (cb {:type         (if (string/starts-with? (:community-id route-params) "z")
```
Comment (ilmotta) [inline src/status_im/router/core.cljs]: I think the `handle-uri` function shouldn't have the responsibility of knowing details such as if the community-id starts with `z`. Example refactor:

```clojure
(defn community-route-type
  [route-params]
  (if (string/starts-with? (:community-id route-params) "z")
    :desktop-community
    :community))
```

File: src/status_im/native_module/core.cljs [hunk 11]
```diff
@@ -278,6 +279,52 @@
...
+
+(defn decompress-key-into-public-key
+  "Provides compressed key to status-go and gets back the uncompressed public key"
+  [public-key callback]
+  (log/info "[native-module] Decompressing compressed key"
+            {:fn         :deserialise-compressed-key-into-public-key
+             :public-key public-key})
+  (.decompressPublicKey ^js (status) public-key callback))
+
+(defn compress-given-key
```
Comment (ilmotta) [inline src/status_im/native_module/core.cljs]: A couple things about this function:

1. Why not just `compress-key`?, The `given` word seems superfluous.
2. The parameter `some-key` uses the word `some`, which is superfluous. Just `key` seems appropriate. `some-`something can be used in almost any parameter anywhere, like `some-callback`, but it's just noise. 
3. The log message strictly says `:public-key`, but the docstring says "Provides ANY key". Also the native call explicitly uses a method named `compressPublickKey`. So is it really *any key* or is it a *public key*? I think we should adjust the code to be clear.

File: src/status_im/native_module/core.cljs [hunk 12]
```diff
@@ -278,6 +279,52 @@
...
+  (let [serialisation-key constants/serialization-key
+        multi-code-prefix constants/multi-code-prefix
+        multi-code-key    (str multi-code-prefix (subs public-key 2))]
+    (log/info "[native-module] Compressing public key"
+              {:fn             :serialise-public-key-into-compressed-key
+               :public-key     public-key
+               :multi-code-key multi-code-key})
+    (.multiformatSerializePublicKey ^js (status) multi-code-key serialisation-key callback)))
+
+(defn deserialise-compressed-key-into-public-key
```
Comment (ilmotta) [inline src/status_im/native_module/core.cljs]: In Clojure there's a very strong convention of naming functions that do type/format conversions using *arrows*. See in the Clojure Style Guide (https://github.com/bbatsov/clojure-style-guide#conversion-functions).

Suggestions:

- Rename `deserialise-compressed-key-into-public-key` to `compressed-key->public-key`
- Rename `serialise-public-key-into-compressed-key` to `public-key->compressed-key`.
- Remove the docstrings from `deserialise-compressed-key-into-public-key` and `serialise-public-key-into-compressed-key` because they seem redundant, especially after the suggested renames. 

Interestingly after renaming we get to see one function is the reverse of the other. Notice I dropped the `serialize`/`deserizalize` words, as this is less important IMO after the renames.

File: src/status_im/native_module/core.cljs [hunk 13]
```diff
@@ -278,6 +279,52 @@
              :connection-string connection-string})
   (.inputConnectionStringForBootstrapping ^js (status) connection-string config-json callback))
 
+(defn deserialize-and-compress-key
```
Comment (ilmotta) [inline src/status_im/native_module/core.cljs:282]: A couple things about this function:

1. If the dosctring says the function receives a `community id public key`, why is the name of the parameter called `desktop-key`?
2. Why is the docstring strictly speaking of a community ID, but the function name is generic? It seems the function works with any key, not just "public community ID keys". If this is the case, then we should try to keep the function agnostic.
3. The log `:fn` doesn't match the actual name of the function.

> gets back a key that status-mobile understands

This seems unnecessary. Everything developed in and for mobile should be understandable for mobile.

Suggestion for change (I removed the docstring because the function is clear enough without one):

```clojure
(defn deserialize-and-compress-key
  [key callback]
  (log/info "[native-module] Compressing public key"
            {:fn          :deserialize-and-compress-key
             :desktop-key key})
  (.deserializeAndCompressKey ^js (status) key callback))
```

File: src/status_im/constants.cljs [hunk 14]
```diff
@@ -211,3 +211,18 @@
...
   "cs")
 
+(def serialization-key
+  "We pass this serialization key as a parameter to MultiformatSerializePublicKey function at status-go,
+  This key determines the output base of the serialisation.
+  according to https://specs.status.im/spec/2#public-key-serialization we serialise keys with base58btc encoding"
+  "z")
+
+(def de-serialization-key
+  "We pass this de-serialization key as a parameter to MultiformatDeserializePublicKey function at status-go,
```
Comment (ilmotta) [inline src/status_im/constants.cljs]: Please, join the dosctring lines accordingly and break them at the 80-100 char length. Same in the var `serialization-key`. See https://github.com/bbatsov/clojure-style-guide#docstring-summary

FYI: See my other comment below about standardizing on a single way to name `deserialize` in English.

File: src/status_im2/navigation/events.cljs [hunk 15]
```diff
@@ -8,7 +8,9 @@
     (all-screens-params (:screen screen-params) (:params screen-params))
 
     (seq screen-params)
-    (assoc-in [:navigation/screen-params view] screen-params)))
+    (assoc-in [:navigation/screen-params view] screen-params)
+  )
```
Comment (flexsurfer) [inline src/status_im2/navigation/events.cljs]: could you rollback this file ?

File: modules/react-native-status/android/src/main/java/im/status/ethereum/module/StatusModule.java [hunk 16]
```diff
@@ -985,6 +985,95 @@ public void run() {
         StatusThreadPoolExecutor.getInstance().execute(runnableTask);
     }
 
+    @ReactMethod
```
Comment (flexsurfer) [inline modules/react-native-status/android/src/main/java/im/status/ethereum/module/StatusModule.java:988]: could you please elaborate on why we need these methods in react ?

File: src/status_im2/navigation/events.cljs (hunk 15 above)
Comment (siddarthkay) [reply src/status_im2/navigation/events.cljs]: @flexsurfer : sure

File: modules/react-native-status/android/src/main/java/im/status/ethereum/module/StatusModule.java (hunk 16 above)
Comment (siddarthkay) [reply modules/react-native-status/android/src/main/java/im/status/ethereum/module/StatusModule.java:988]: @flexsurfer : we don't actually need the first 3 methods yet only the last one : `deserializeAndCompressKey` 
These got left in because I was experimenting with them to see which one of these methods were able to de compress the community id we receive from desktop to make it into a format we see on mobile..

is okay to keep the unused methods for future or I could delete them now?

File: src/status_im/ui/screens/chat/message/link_preview.cljs (hunk 8 above)
Comment (siddarthkay) [reply src/status_im/ui/screens/chat/message/link_preview.cljs]: hmm yeah I will revert the changes here, I was doing some work here but then decided to revert it, which left behind this code block.

File: src/status_im/utils/universal_links/core.cljs (hunk 6 above)
Comment (siddarthkay) [reply src/status_im/utils/universal_links/core.cljs]: Okay I will make it `deserialize` since that is how its present in status-go. more consistency :)

File: src/status_im/ui/screens/chat/message/link_preview.cljs (hunk 9 above)
Comment (siddarthkay) [reply src/status_im/ui/screens/chat/message/link_preview.cljs]: will do :)

File: src/status_im/native_module/core.cljs (hunk 13 above)
Comment (siddarthkay) [reply src/status_im/native_module/core.cljs:282]: > This seems unnecessary. Everything developed in and for mobile should be understandable for mobile.

@ilmotta 
Hmm I wanted to be specific here because previously in certain scenarios we'd have community keys in chat which status-mobile would not understand and hence the "community not found" errors would popup. This PR was made to fix that.. and hence the docstring.. 
Does this make sense now?

File: src/status_im/router/core.cljs (hunk 10 above)
Comment (siddarthkay) [reply src/status_im/router/core.cljs]: Good suggestion! Thanks @ilmotta

File: src/status_im/native_module/core.cljs (hunk 12 above)
Comment (siddarthkay) [reply src/status_im/native_module/core.cljs]: The method `deserialise-compressed-key-into-public-key` explains that we are achieving compressed-key->public-key by deserialization. renaming it to `compressed-key->public-key` will mean we are getting a->b , but we will lose the how.. should we then keep the docstrings and rename as you suggested ? what do you think ?

File: src/status_im/native_module/core.cljs (hunk 13 above)
Comment (ilmotta) [reply src/status_im/native_module/core.cljs:282]: I understand the reasoning @siddarthkay, but to me this is still complecting domain-related stuff in a generic implementation. This is truly a utility and wrapper function, agnostic of domain.

> previously in certain scenarios we'd have community keys in chat which status-mobile would not understand and hence the "community not found" errors would popup.

The onus should be on the calling side. This function is agnostic of which key it receives, so the caller is responsible for using it appropriately, handling errors, etc.

---

I'm still confused by the parameter name `desktop-key`. What's a desktop key? Could it be just `key`?

File: src/status_im/native_module/core.cljs (hunk 12 above)
Comment (ilmotta) [reply src/status_im/native_module/core.cljs]: Great question! In general, I think function names should never explain the *how*, only the *what*. The *how* is always on the implementation side or in the docstrings/comments. 

> should we then keep the docstrings and rename as you suggested

Precisely @siddarthkay, if you feel like it's important to give more information about *how* something works that's not clear by the code itself, then a comment/docstring works wonders.

File: src/status_im/native_module/core.cljs (hunk 13 above)
Comment (siddarthkay) [reply src/status_im/native_module/core.cljs:282]: well.. naming variables / parameters is indeed one the things i suck at :)
my thought process behind keeping the parameter `desktop-key` was to remind me that here goes the community-id (key) we get from desktop and in callback we would get the key that mobile understands. ( for me .. `mobile-key`)
But as I understand this causes confusion amongst other devs who do not know the context, I'll be happy to rename it to be just `key`

Comment (ilmotta) [reply src/status_im/native_module/core.cljs:282]: I see your point. It might be worth a shot adding these details to the docstring, but it's sure difficult for people without context to do this connection, given how generic this function is.

Comment (siddarthkay) [reply src/status_im/native_module/core.cljs:282]:
```clojure
(defn deserialize-and-compress-key
  "Provides a community id (public key) to status-go which is first deserialized
  and then compressed. Example input/output :
  input key  = zQ3shTAten2v9CwyQD1Kc7VXAqNPDcHZAMsfbLHCZEx6nFqk9 and
  output key = 0x025596a7ff87da36860a84b0908191ce60a504afc94aac93c1abd774f182967ce6"
  [key callback]
  (log/info "[native-module] Deserializing and then compressing public key"
            {:fn  :deserialize-and-compress-key
             :key key})
  (.deserializeAndCompressKey ^js (status) key callback))
```
  
  @ilmotta : this better?

File: modules/react-native-status/android/src/main/java/im/status/ethereum/module/StatusModule.java (hunk 16 above)
Comment (siddarthkay) [reply modules/react-native-status/android/src/main/java/im/status/ethereum/module/StatusModule.java:988]: as far as I know Erik needs one of these methods in his PR

File: src/status_im/native_module/core.cljs (hunk 13 above)
Comment (ilmotta) [reply src/status_im/native_module/core.cljs:282]: For sure @siddarthkay :) Thanks! Don't kill me, but if you want to go the extra mile a test is like *executable documentation*, so we could have a simple test describing the input/output.

Comment (siddarthkay) [reply src/status_im/native_module/core.cljs:282]: A test for this exists in status-go and we are just calling this method in status-mobile :) 
https://github.com/status-im/status-go/pull/3042/files#diff-79a6d848d7e638385a6ebc36d289250f869979e18cd17c9c1ef810da690836cfL27

Comment (ilmotta) [reply src/status_im/native_module/core.cljs:282]: Sounds good, given this function is just a wrapper. Thanks!

File: modules/react-native-status/android/src/main/java/im/status/ethereum/module/StatusModule.java [hunk 17]
```diff
@@ -985,6 +985,95 @@ public void run() {
...
+                String res = Statusgo.multiformatDeserializePublicKey(multiCodecKey,base58btc);
+                callback.invoke(res);
+            }
+        };
+
+        StatusThreadPoolExecutor.getInstance().execute(runnableTask);
+    }
+
+    @ReactMethod
+    public void compressPublicKey(final String multiCodecKey, final Callback callback) throws JSONException {
```
Comment (ilmotta) [inline modules/react-native-status/android/src/main/java/im/status/ethereum/module/StatusModule.java:1025]: Just throwing a comment here to raise awareness @siddarthkay, but there's nothing preventing us from better modularizing the file `StatusModule.java` and using it only to expose public methods, but not the actual implementation, which would live in separate classes however we see fit. This separation would further help writing tests for these methods, which are dearly missed.

Currently this file is basically a kitchen sink for random bits of Java code.

File: src/status_im/utils/universal_links/core.cljs [hunk 18]
```diff
@@ -76,7 +77,24 @@
...
+  (navigation/navigate-to-cofx cofx :community {:community-id deserialized-key})
+)
+
+(rf/defn handle-desktop-community
+  [cofx {:keys [community-id]}]
+  (status/deserialize-and-compress-key
+   community-id
+   (fn [deserialized-key]
+     (rf/dispatch [:handle-navigation-to-desktop-community-from-mobile cofx (str deserialized-key)])
+   )
```
Comment (ilmotta) [inline src/status_im/utils/universal_links/core.cljs]: There are some dangling parentheses in this file.

**Review by ilmotta — APPROVED** [review]

File: modules/react-native-status/android/src/main/java/im/status/ethereum/module/StatusModule.java [hunk 19]
```diff
@@ -985,6 +985,95 @@ public void run() {
...
+                String res = Statusgo.multiformatSerializePublicKey(multiCodecKey,base58btc);
+                callback.invoke(res);
+            }
+        };
+
+        StatusThreadPoolExecutor.getInstance().execute(runnableTask);
+    }
+
+    @ReactMethod
+    public void multiformatDeserializePublicKey(final String multiCodecKey, final String base58btc, final Callback callback) throws JSONException {
```
Comment (ilmotta) [inline modules/react-native-status/android/src/main/java/im/status/ethereum/module/StatusModule.java:1007]: Not a suggestion for this PR, just an idea.

There are a bunch of methods here which require the same plumbing to execute the body of `run` in the thread pool. We could easily create a method that receives a lambda to remove all the duplication.

Comment (siddarthkay) [reply modules/react-native-status/android/src/main/java/im/status/ethereum/module/StatusModule.java:1007]: yeah this is a good idea, I'll do this in a separate PR.
Link to issue : https://github.com/status-im/status-mobile/issues/14687

File: modules/react-native-status/android/src/main/java/im/status/ethereum/module/StatusModule.java (hunk 17 above)
Comment (siddarthkay) [reply modules/react-native-status/android/src/main/java/im/status/ethereum/module/StatusModule.java:1025]: Hmm, this is worth doing a quick experiment on, I will log my findings here : https://github.com/status-im/status-mobile/issues/14688
Would love to modularize further  🫶🏻

File: src/status_im/ui/screens/chat/message/link_preview.cljs (hunk 9 above)
Comment (siddarthkay) [reply src/status_im/ui/screens/chat/message/link_preview.cljs]: here is the issue to track broke community link previews : https://github.com/status-im/status-mobile/issues/14689

File: src/status_im/utils/universal_links/core.cljs (hunk 18 above)
Comment (siddarthkay) [reply src/status_im/utils/universal_links/core.cljs]: I wish these were caught with z-print or something, for the time being i will stay vigilant

File: src/status_im/constants.cljs [hunk 20]
```diff
@@ -211,3 +211,20 @@
   An example of a connection string is -> cs2:5vd6J6:Jfc:27xMmHKEYwzRGXcvTtuiLZFfXscMx4Mz8d9wEHUxDj4p7:EG7Z13QScfWBJNJ5cprszzDQ5fBVsYMirXo8MaQFJvpF:3 "
   "cs")
 
+(def serialization-key
```
Comment (qfrank) [inline src/status_im/constants.cljs]: can we add ^:const before symbol? @siddarthkay

Comment (siddarthkay) [reply src/status_im/constants.cljs]: although variables in clojure are immutable, I'll go ahead and add the ^:const metadata

**Review by qfrank — APPROVED** [review]

**Comment by siddarthkay** [comment]
@status-im/mobile-qa : This PR gives us the ability to join a community which was made from Desktop.
However link previews do not work for any community ( even ones that were made from mobile), it is being tracked here : https://github.com/status-im/status-mobile/issues/14689

File: modules/react-native-status/android/src/main/java/im/status/ethereum/module/StatusModule.java (hunk 16 above)
Comment (erikseppanen) [reply modules/react-native-status/android/src/main/java/im/status/ethereum/module/StatusModule.java:988]: yes @siddarthkay I'll need to convert "z..." compressed keys to uncompressed "0x..." keys. thanks!

**Comment by VolodLytvynenko** [comment]
hi @siddarthkay  Thank you for PR. the ISSUE 1 https://github.com/status-im/status-mobile/issues/14560 is still reproducible

 **ISSUE 1 "Not found" screen when following a link to a community created on Desktop**

**Steps to reproduce are the same**:
1. User A (desktop) adds User B (mobile) to contacts
2. User A: create a community on desktop
3. User A: go to Overview > Share community > Copy the link
4. User A: paste the link in 1-1 chat with User B and send the message
5. User B: follow the link

**Actual result:**
- The link is not resolved
- Redirect to the Not found screen
https://user-images.githubusercontent.com/52490791/210395118-1bf47330-e8e7-42ff-906b-8ac9ddd1d392.mp4

**Expected result:**
- The link is resolved in the 1-1 chat view as an invitation link (with the View button)
- Redirect to the community when following the link

**Additional Information**
- Operating System: Android, iOS

**Comment by siddarthkay** [comment]
> hi @siddarthkay Thank you for PR. the ISSUE 1 #14560 is still reproducible
> 
> **ISSUE 1 "Not found" screen when following a link to a community created on Desktop**
> 
> **Steps to reproduce are the same**:
> 
> 1. User A (desktop) adds User B (mobile) to contacts
> 2. User A: create a community on desktop
> 3. User A: go to Overview > Share community > Copy the link
> 4. User A: paste the link in 1-1 chat with User B and send the message
> 5. User B: follow the link
> 
> **Actual result:**
> 
> * The link is not resolved
> * Redirect to the Not found screen
>   https://user-images.githubusercontent.com/52490791/210395118-1bf47330-e8e7-42ff-906b-8ac9ddd1d392.mp4
> 
> **Expected result:**
> 
> * The link is resolved in the 1-1 chat view as an invitation link (with the View button)
> * Redirect to the community when following the link
> 
> **Additional Information**
> 
> * Operating System: Android, iOS

Hi @VladimrLitvinenko : just for my curiosity could you let me know what happens when you tap on "fetch community" button in this case?

**Comment by siddarthkay** [comment]
@VladimrLitvinenko : 
Responding to Expected result section of ISSUE 1:
you will also not be seeing  "view community" button when you paste the link, as mentioned in this PR above, that issue with previews is being separately tracked here https://github.com/status-im/status-mobile/issues/14689

**Review by yqrashawn — APPROVED** [review]
Thanks for your contribution. LGTM! :thumbsup:

File: status-go-version.json (hunk 1 above)
Comment (siddarthkay) [reply status-go-version.json]: updated status-go version to sha1 of latest commit in status-go

File: modules/react-native-status/android/src/main/java/im/status/ethereum/module/StatusModule.java (hunk 16 above)
Comment (flexsurfer) [reply modules/react-native-status/android/src/main/java/im/status/ethereum/module/StatusModule.java:988]: could you please elaborate on why do we need to do this in react ?

Comment (siddarthkay) [reply modules/react-native-status/android/src/main/java/im/status/ethereum/module/StatusModule.java:988]: Sure, @flexsurfer 
lets say you make a new community in desktop 
and you click on share and you get a link like this : https://join.status.im/c/zQ3shTAten2v9CwyQD1Kc7VXAqNPDcHZAMsfbLHCZEx6nFqk9
Then when you paste this link in chat on mobile to another user,
Our existing react code picks up zQ3shTAten2v9CwyQD1Kc7VXAqNPDcHZAMsfbLHCZEx6nFqk9  as the community id and searches for it in the mailserver ( if it is not already a part of our db ).

This community is not found because zQ3shTAten2v9CwyQD1Kc7VXAqNPDcHZAMsfbLHCZEx6nFqk9  is a short key ( some people call it compressed key or serialized key )

hence we need to convert it to a community id that can be searchable which begins with 0x02 format.
To achieve this we need to pass this short key to status go and get back the format which allows us to discover and join communities.

Comment (flexsurfer) [reply modules/react-native-status/android/src/main/java/im/status/ethereum/module/StatusModule.java:988]: that sounds strange,`that can be searchable which begins with 0x02 format.` why can't it be searchable with the key from desktop? where do we search ?

Comment (siddarthkay) [reply modules/react-native-status/android/src/main/java/im/status/ethereum/module/StatusModule.java:988]: I do not know why things are the way they are, but the search code is here: https://github.com/status-im/status-go/pull/2879#discussion_r1058222003

Comment (siddarthkay) [reply modules/react-native-status/android/src/main/java/im/status/ethereum/module/StatusModule.java:988]: it would be ideal if all the keys followed same formats and were consistent, but my hypothesis is that status-desktop team updates the format of the key on the client side to make it appear shorter, else they should have been searchable by the mailserver.

If that was not the case then we would have never received this issue : https://github.com/status-im/status-mobile/issues/14560

Comment (siddarthkay) [reply modules/react-native-status/android/src/main/java/im/status/ethereum/module/StatusModule.java:988]: @flexsurfer : Desktop does this on client side : https://github.com/status-im/status-desktop/pull/5269

Comment (flexsurfer) [reply modules/react-native-status/android/src/main/java/im/status/ethereum/module/StatusModule.java:988]: omg why do we need this headache, short keys have zero sense :( i already had lots of issues restoring my accounts from a mobile on desktop, I know its a rhetorical question, *just crying in the corner*

**Comment by VolodLytvynenko** [comment]
@siddarthkay 
> Responding to Expected result section of ISSUE 1:
you will also not be seeing "view community" button when you paste the link, as mentioned in this PR above, that issue with previews is being separately tracked here https://github.com/status-im/status-mobile/issues/14689

Oh, I didn't understand that it is about this preview page. That's ok If it'll be fixed separately

> Hi @VladimrLitvinenko : just for my curiosity could you let me know what happens when you tap on "fetch community" button in this case?

The community page is opened then after  "fetch community"  button is tapped. But initially (when the link first time is tapped) the not found page is shown. Is it expected behavior?
All other work correctly and can be merged

**Comment by siddarthkay** [comment]
> The community page is opened then after "fetch community" button is tapped. But initially (when the link first time is tapped) the not found page is shown. Is it expected behavior?

This may or may not be further improved but at the moment this is expected, reason is that the new community is not part of internal Db since it was newly created and when you press the "fetch" button it is then retrieved from the mailserver.

**Comment by siddarthkay** [comment]
> This may or may not be further improved but at the moment this is expected, reason is that the new community is not part of internal Db since it was newly created and when you press the "fetch" button it is then retrieved from the mailserver.

I'll make a separate research issue on GitHub for this and see if it is possible to improve this UX further :)

File: modules/react-native-status/android/src/main/java/im/status/ethereum/module/StatusModule.java (hunk 16 above)
Comment (siddarthkay) [reply modules/react-native-status/android/src/main/java/im/status/ethereum/module/StatusModule.java:988]: Well this PR just makes sure that we are able to understand short keys, i.e join a community which was made by desktop app. 
This PR does not introduce short key format for communities in mobile UI. 

Also this capability will be essential when we do RC1 release and mobile users try to join the dog-fooding communities made by Desktop :)

**Comment by siddarthkay** [comment]
@VladimrLitvinenko @status-im/mobile-qa  : so is this PR good to be merged or does this need to be moved to E2E?

**Comment by churik** [comment]
Best to wait for e2e, thanks Sid! (I've already launched them, so waiting for results and ready to go)

**Review by erikseppanen — APPROVED** [review]

**Comment by churik** [comment]
ready to go!
